controllers/managedclusterForCGU_controller.go

```diff
@@ -0,0 +1,207 @@
+}
+
+// Reconcile the managed cluster auto create ClusterGroupUpgrade
+// - When a new managed cluster is created, create ClusterGroupUpgrade CR for the cluster only when it's ready
+//   and its child policies are available
+//
+// Note: The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ManagedClusterForCguReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("Request.Name", req.Name)
+	reqLogger.Info("Reconciling managedCluster to create clusterGroupUpgrade")
+
+	managedCluster := &clusterv1.ManagedCluster{}
+	err := r.Get(ctx, types.NamespacedName{Name: req.Name}, managedCluster)
+	if errors.IsNotFound(err) {
+		// managed cluster could have been deleted
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		// Error reading the object, requeue the request
+		return ctrl.Result{}, err
+	}
+
+	clusterGroupUpgrade := &ranv1alpha1.ClusterGroupUpgrade{}
+	err = r.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: ztpInstallNS}, clusterGroupUpgrade)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading clusterGroupUpgrade, requeue the request
+			return ctrl.Result{}, err
+		}
+	} else {
+		// clusterGroupUpgrade for this cluster already exists, stop reconcile
+		r.Log.Info("clusterGroupUpgrade found", "Name", clusterGroupUpgrade.Name, "Namespace", clusterGroupUpgrade.Namespace)
+		return ctrl.Result{}, nil
+	}
+
+	availableCondition := meta.FindStatusCondition(managedCluster.Status.Conditions, clusterv1.ManagedClusterConditionAvailable)
+	if availableCondition == nil {
+		r.Log.Info("cluster has no status yet", "Name", managedCluster.Name, "RequeueAfter:", clusterNoStatusRetryDelay)
+		return ctrl.Result{RequeueAfter: clusterNoStatusRetryDelay}, nil
+	} else if availableCondition.Status == metav1.ConditionTrue {
+		// cluster is ready
+		policies, err := r.getPolicies(ctx, managedCluster.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if policies == nil {
+			// child policies have not been created, requeue the request
+			return ctrl.Result{RequeueAfter: clusterNotReadyRetryDelay}, nil
+		}
+
+		// create clusterGroupUpgrade
+		if err = r.newClusterGroupUpgrade(ctx, managedCluster, policies); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		// availableCondition is false or unknown, cluster is not ready
+		r.Log.Info("cluster is not ready", "RequeueAfter:", clusterNotReadyRetryDelay)
+		return ctrl.Result{RequeueAfter: clusterNotReadyRetryDelay}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ManagedClusterForCguReconciler) getPolicies(ctx context.Context, clusterName string) (*policiesv1.PolicyList, error) {
+	policies := &policiesv1.PolicyList{}
+	if err := r.List(ctx, policies, client.InNamespace(clusterName)); err != nil {
+		return nil, err
+	}
+	return policies, nil
```

Comment:
I'm wondering if we need an extra check to make sure these policies are indeed child policies and not another policy that somehow ended up in the same namespace? Child policies have a specific label.

Reply:
From [ACM doc](https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/governance/governance#policy-overview), it says "You can create a policy in any namespace on the hub cluster except the cluster namespace. If you create a policy in the cluster namespace, it is deleted by Red Hat Advanced Cluster Management for Kubernetes."

And I tried to create policies under cluster namespace, it was deleted like the doc describes. We should be safe.